components/ai_chat/core/browser/engine/oai_parsing.cc

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/ai_chat/core/browser/engine/oai_parsing.h"
+
+#include <string>
+#include <utility>
+
+#include "base/json/json_reader.h"
+#include "brave/components/ai_chat/core/browser/engine/engine_consumer.h"
+#include "brave/components/ai_chat/core/browser/tools/tool.h"
+
+namespace ai_chat {
+
+namespace {
+
+// Prevent parsing arguments list that's too large for the
+// tools we're providing.
+const size_t kMaxArgumentsLength = 1000000;
+
+// Prevent hallucinated JSON structures
+const size_t kMaxArgumentsJSONDepth = 20;
+
+}  // namespace
+
+std::vector<mojom::ToolUseEventPtr> ToolUseEventFromToolCallsResponse(
+    const base::Value::List* tool_calls_api_response) {
+  // https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools
+  // https://platform.openai.com/docs/api-reference/chat/object
+  // choices -> message -> tool_calls
+  std::vector<mojom::ToolUseEventPtr> tool_use_events;
+  for (auto& tool_call_raw : *tool_calls_api_response) {
+    if (!tool_call_raw.is_dict()) {
+      DLOG(ERROR) << "Tool call is not a dictionary.";
+      continue;
+    }
+    const auto& tool_call = tool_call_raw.GetDict();
+    mojom::ToolUseEventPtr tool_use_event = mojom::ToolUseEvent::New();
+
+    const base::Value::Dict* function = tool_call.FindDict("function");
+    if (!function) {
+      DLOG(ERROR) << "No function info found in tool call.";
+      continue;
+    }
+    // Tool call results can be partial and should be added to the previous
+    // event by the event handler.
+    const std::string* id = tool_call.FindString("id");
+    if (!id) {
+      DLOG(ERROR) << "No id found in tool call.";
+      continue;
+    }
```

Comment:
`mojom::ToolUseEvent::New()` can be put after we are sure it contains `function` and `id`